ree/reeCorrectedFlux.cc

```diff
@@ -0,0 +1,299 @@
+/*
+ * Copyright (C) 2011 Scientific Computation Research Center
+ *
+ * This work is open source software, licensed under the terms of the
+ * BSD license as described in the LICENSE file in the top-level directory.
+ */
+#include <apfCavityOp.h>
+#include "apfElement.h"
+#include "crv.h"
+#include "crvShape.h"
+#include "ree.h"
+
+namespace ree {
+
+enum {VISITED};
+
+/* overall information useful during computation of corrected flux */
+struct CorrectFlux
+{
+  apf::Mesh* mesh;
+  /* mesh dimension, so far handling 3 only */
+  int dim;
+  /* polynomial order of nedelec space */
+  int order;
+  /* polynomial order of p+1 nedelec space */
+  int orderp1;
+  /* input nedelec field containing scalar dofs for electric field */
+  apf::Field* ef;
+  /* input theta field for flux correction
+   * (3 scalar dofs on each face, 1 per edge) */
+  apf::Field* theta;
+  /* tags each face once it has been visited */
+  apf::MeshTag* tag;
+  /* output per-element field containing correctedFlux vectors,
+   * which are computed using theta and electric fields */
+  apf::Field* correctedFlux;
+};
+
+static void setupCorrectFlux(
+    CorrectFlux* cf,
+    apf::Field* f,
+    apf::Field* theta,
+    apf::Field* correctedFlux)
+{
+  cf->mesh = apf::getMesh(f);
+  cf->dim = cf->mesh->getDimension();
+  cf->order = f->getShape()->getOrder();
+  cf->orderp1 = cf->order+1;
+  cf->ef = f;
+  cf->theta = theta;
+  cf->tag = cf->mesh->createIntTag("isVisited", 1);
+
+  cf->correctedFlux = correctedFlux;
+  int nc = apf::countComponents(cf->correctedFlux);
+  double zeros[nc];
+  for (int i = 0; i < nc; i++)
+    zeros[i] = 0.;
+  apf::MeshEntity* tet;
+  apf::MeshIterator* it = cf->mesh->begin(3);
+  while ((tet = cf->mesh->iterate(it))) {
+    apf::MeshElement* me = apf::createMeshElement(cf->mesh, tet);
+    int orderp1 = cf->order+1;
+    int np = apf::countIntPoints(me, 2*orderp1-1);
+
+    for (int i = 0; i < np; i++) {
+      apf::setComponents(cf->correctedFlux, tet, i, zeros);
+    }
```

Comment:
@Samiullah-Malik 

there needs to be a `apf::destroyMeshElement(me)` call after this line within the while loop.

Reply:
Done.